src/Microsoft.Data.SqlClient/netcore/src/Microsoft/Data/SqlClientX/IO/TdsReadStream.cs

```diff
@@ -0,0 +1,341 @@
+                    _readIndex = 0;
+                }
+                // 1.2. There is no data left as indicated by packet header and the buffer is empty.
+                else if (_packetDataLeft == 0)
+                {
+                    _readBufferDataEnd = isAsync ?
+                        await _underlyingStream.ReadAsync(_readBuffer, ct).ConfigureAwait(false) :
+                        _underlyingStream.Read(_readBuffer);
+                    
+                    _readIndex = 0;
+
+                    await ProcessHeaderAsync(isAsync, ct).ConfigureAwait(false);
+
+                    // 1.3. After processing the packet header, there is a possibility that the transport read didn't
+                    // return any more data for the packet. In that case, post another read to have packet data ready..
+                    if (_readBufferDataEnd == _readIndex)
+                    {
+                        _readBufferDataEnd = isAsync ?
+                            await _underlyingStream.ReadAsync(_readBuffer, ct).ConfigureAwait(false) :
+                            _underlyingStream.Read(_readBuffer);
+                    }
+                }
+            }
+        }
+
+        /// <summary>
+        /// If the buffer is empty, or the packet data is completely read, 
+        /// then we need to prepare the buffer for the next packet.
+        /// </summary>
+        /// <param name="isAsync"></param>
+        /// <param name="ct"></param>
+        /// <returns></returns>
+        private async ValueTask PrepareBufferIfNeeded(bool isAsync, CancellationToken ct)
+        {
+            // If there we have read through the packet, 
+            // or if we have reached the end of the buffer,
+            // then we need to position the buffer at the beginning of packet or buffer
+            if (_packetDataLeft == 0 || _readBufferDataEnd == _readIndex)
+                await PrepareBufferAsync(isAsync, ct).ConfigureAwait(false);
+        }
+
+        /// <summary>
+        /// Processes the header of the packet, and extracts the data from the header.
+        /// If needed, this function will read more data from the network to complete the header.
+        /// </summary>
+        /// <param name="isAsync">Whether this method is invoked asynchronously.</param>
+        /// <param name="ct">The cancellation token.</param>
+        /// <returns></returns>
+        private async ValueTask ProcessHeaderAsync(bool isAsync, CancellationToken ct)
+        {
+            int headerDataAvailable = _readBufferDataEnd - _readIndex;
+            int bytesNeededToCompleteHeader = TdsEnums.HEADER_LEN - headerDataAvailable;
+
+            // We have less than the header length available in the buffer, so we need to read more data, to atleast complete 
+            // the header.
+            if (headerDataAvailable < TdsEnums.HEADER_LEN)
+            {
+                // We move the header information to the beginning of the buffer.
```

Comment:
Do we need to do this, or can we read to index zero in `_readBuffer` from the outset?

Reply:
This is a special case, where a MITM is splitting the packets (inspired by the comments in the current sqlclient implementation), and we get a part of the next packet while reading an existing packet, but didnt get the full header. so we are talking about less than 8 bytes of header in the buffer.
At this point we need the full header, and allow as much space in the buffer to accomodate the full TDS packet.

Reply:
Going to resolve this conversation. But feel free to reopen if I am missing something.